File: 2020/Day8/Python/HandheldHalting-Part2.py
import sys
import collections
import itertools
import numpy as np

def run(code):
    acc = 0
    instruction = 0
    seen = []
    while instruction not in seen:
        seen.append(instruction)
        
        if instruction >= len(code):
            return True, acc
        op, arg = code[instruction]
        if op == "nop":
            instruction += 1
        elif op == "acc":
            acc += arg
            instruction += 1
        elif op == "jmp":
            instruction += arg
    return False, acc

code = []

for line in open(f"{sys.path[0]}/../input.txt").read().splitlines():
    op, arg = line.split()
    code.append((op, int(arg)))

instruction = 0
while True:
    op, arg = code[instruction]
    x = code.copy()

    if op == "acc":
        instruction += 1

    elif op == "nop":
        x[instruction] = ("jmp", arg)
        valid, ans = run(x)
        instruction += 1

    elif op == "jmp":
        x[instruction] = ("nop", arg)
        instruction += arg

    valid, ans = run(x)
    if valid:
        print(ans)
        break

File: 2020/Day8/Haskell/HandheldHalting-Part2.hs
import System.IO
import Control.Monad

printArray:: (Show a, Eq a) => [a] -> String
printArray (x:xs)
    | null xs = show x
    | otherwise = show x ++ "\n" ++ printArray xs

strip:: (Eq a) => [a] -> a -> [a]
strip [] _ = []
strip (x:xs) a
    | x == a = strip xs a
    | otherwise = x : strip xs a

runCode:: Int -> Int -> [Int] -> [([Char], Int)] -> (Bool, Int)
runCode acc ins seen code 
    | ins `elem` seen = (False, acc)
    | length code <= ins = (True, acc)
    | op == "nop" = runCode acc (ins+1) (ins:seen) code
    | op == "jmp" = runCode acc (ins+arg) (ins:seen) code
    | op == "acc" = runCode (acc+arg) (ins+1) (ins:seen) code
    where (op, arg) = code!!ins 

fixCode:: Int -> [([Char], Int)] -> Int
fixCode ins code 
    | valid = ans
    | op == "nop" = fixCode (ins+1) code
    | op == "jmp" = fixCode (ins+arg) code
    | op == "acc" = fixCode (ins+1) code
    where 
        (op, arg) = code!!ins
        (valid, ans) = runCode 0 0 [] (replaceInstruction ins code)

replaceInstruction:: Int -> [([Char], Int)] -> [([Char], Int)]
replaceInstruction ins code
    | op == "nop" = take ins code ++ [("jmp", arg)] ++ drop (ins+1) code
    | op == "jmp" = take ins code ++ [("nop", arg)] ++ drop (ins+1) code
    | otherwise = code
    where (op, arg) = code!!ins

parseInstruction:: [Char] -> ([Char], Int)
parseInstruction ins = (op, read arg::Int)
    where (op:arg:_) = split ' ' (strip ins '+')

main:: IO()
main = do
    handle <- openFile "../input.txt" ReadMode
    contents <- hGetContents handle
    let rows = lines contents
    print (fixCode 0 (map parseInstruction rows))
    hClose handle